v1/plugin/plugin.go

```diff
@@ -312,29 +507,212 @@ func startPlugin(srv server, m *meta, p *pluginProxy) int {
 	addr := fmt.Sprintf("127.0.0.1:%v", l.Addr().(*net.TCPAddr).Port)
 	lis, err := net.Listen("tcp", addr)
 	if err != nil {
-		// TODO(danielscottt): logging
-		panic(err)
+		return "", err
 	}
 	go func() {
 		err := srv.Serve(lis)
 		if err != nil {
-			panic(err)
+			log.Fatal(err)
 		}
 	}()
+	pprofAddr := "0"
+	if isPprof {
+		pprofAddr, err = startPprof()
+		if err != nil {
+			return "", err
+		}
+	}
 	resp := preamble{
 		Meta:          *m,
 		ListenAddress: addr,
 		Type:          m.Type,
-		PprofAddress:  pprofPort,
+		PprofAddress:  pprofAddr,
 		State:         0, // Hardcode success since panics on err
 	}
 	preambleJSON, err := json.Marshal(resp)
 	if err != nil {
 		panic(err)
 	}
-	libInputOutput.printOut(string(preambleJSON))
-	go p.HeartbeatWatch()
-	// TODO(danielscottt): exit code
-	<-p.halt
-	return 0
+
+	return string(preambleJSON), nil
+}
+
+// GetPluginType converts a pluginType to a string
+// describing what type of plugin this is
+func getPluginType(plType pluginType) string {
+	switch plType {
+	case collectorType:
+		return "collector"
+
+	case publisherType:
+		return "publisher"
+
+	case processorType:
+		return "processor"
+	}
+	return ""
+}
+
+// GetRPCType converts a RPCType (int) to a string
+// as described in snap/control/plugin/plugin.go
+func getRPCType(i int) string {
+	if i == 0 {
+		return "NativeRPC (0)"
+	} else if i == 2 {
+		return "GRPC (2)"
+	}
+	return "Not a valid RPC Type"
+}
+
+func showDiagnostics(m meta, p *pluginProxy, c Config) error {
+	defer timeTrack(time.Now(), "showDiagnostics")
+	printRuntimeDetails(m)
+	err := printConfigPolicy(p, c)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	met, err := printMetricTypes(p, c)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	err = printCollectMetrics(p, met)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	printContactUs()
+	return nil
+
+}
+
+func printMetricTypes(p *pluginProxy, conf Config) ([]Metric, error) {
+	defer timeTrack(time.Now(), "printMetricTypes")
+	met, err := p.plugin.(Collector).GetMetricTypes(conf)
+	if err != nil {
+		return nil, fmt.Errorf("! Error in the call to GetMetricTypes: \n%v", err)
+	}
+	//apply any config passed in to met so that
+	//CollectMetrics can see the config for each metric
+	for i := range met {
+		met[i].Config = conf
+	}
+
+	fmt.Println("Metric catalog will be updated to include: ")
+	for _, j := range met {
+		fmt.Printf("    Namespace: %v \n", j.Namespace.String())
+	}
+	return met, nil
+}
+
+func printConfigPolicy(p *pluginProxy, conf Config) error {
```

Comment:
@kjlyon, I know that `printConfigPolicy` is rather not complicated function especially when you acquainted with printed output, however, it would be nice to improve readability of this code inside. My suggestion is split this one to `printConfigPolicyStringRules`, `printConfigPolicyIntegerRules`, etc. as the first step. 
Do you have any other idea what might be done?